mlfinlab/bet_sizing/ef3m.py

```diff
@@ -0,0 +1,394 @@
+        :return: (list) List of the first five moments
+        """
+        u_1, u_2, s_1, s_2, p_1 = parameters  # Expanded mixture parameters to individual variables for clarity.
+        p_2 = 1 - p_1  # Explicitly state p_2 for symmetry.
+        m_1 = p_1 * u_1 + p_2 * u_2  # Eq. (6)
+        m_2 = p_1 * (s_1**2 + u_1**2) + p_2 * (s_2**2 + u_2**2)  # Eq. (7)
+        m_3 = p_1 * (3 * s_1**2 * u_1 + u_1**3) + p_2 * (3 * s_2**2 * u_2 + u_2**3)  # Eq. (8)
+        # Eq. (9)
+        m_4 = p_1 * (3 * s_1**4 + 6 * s_1**2 * u_1**2 + u_1**4) + p_2 * (3 * s_2**4 + 6 * s_2**2 * u_2**2 + u_2**4)
+        # Eq (10)
+        m_5 = p_1 * (15 * s_1**4 * u_1 + 10 * s_1**2 * u_1**3 + u_1**5) + p_2 *\
+            (15 * s_2**4 * u_2 + 10 * s_2**2 * u_2**3 + u_2**5)
+
+        if return_result:
+            return [m_1, m_2, m_3, m_4, m_5]
+
+        self.new_moments = [m_1, m_2, m_3, m_4, m_5]
+        return None
+
+    def iter_4(self, mu_2, p_1):
+        """
+        Evaluation of the set of equations that make up variant #1 of the EF3M algorithm (fitting using the first
+        four moments).
+
+        :param mu_2: (float) Initial parameter value for mu_2
+        :param p_1: (float) Probability defining the mixture; p_1, 1 - p_1
+        :return: (list) List of estimated parameter if no invalid values are encountered (e.g. complex values,
+         divide-by-zero), otherwise an empty list is returned.
+        """
+        m_1, m_2, m_3, m_4 = self.moments[0:4]  # Expand list of moments to individual variables for clarity.
+        param_list = []
+
+        # Using a while-loop here to be able to use 'break' functionality.
+        # We need to stop the calculation at any given step to avoid throwing warnings or errors,
+        # and be in control of our return values. I'm open to other suggestions, but multiple return statements isn't
+        # one of them.
+        while True:
+            # Calculate mu_1, Equation (22).
+            mu_1 = (m_1 - (1 - p_1) * mu_2) / p_1
+
+            # Calculate sigma_2, Equation (24)
+            if (3 * (1 - p_1) * (mu_2 - mu_1)) == 0:
+                # Validity check 1: Check for divide-by-zero.
+                break
+
+            sigma_2_squared = ((m_3 + 2 * p_1 * mu_1**3 + (p_1 - 1) * mu_2**3 - 3 * mu_1 * (m_2 + mu_2**2 * (p_1-1))) /
+                               (3*(1-p_1)*(mu_2-mu_1)))
+            if sigma_2_squared < 0:
+                # Validity check 2: Prevent potential complex values.
+                break
+
+            sigma_2 = sigma_2_squared**0.5
+            # Calculate sigma_1, Equation (23)
+            sigma_1_squared = ((m_2 - sigma_2**2 - mu_2**2) / p_1 + sigma_2**2 + mu_2**2 - mu_1**2)
+
+            if sigma_1_squared < 0:
+                # Validity check 3: Prevent potential complex values.
+                break
+            sigma_1 = sigma_1_squared**0.5
+
+            #if np.iscomplex(sigma_1) or np.iscomplex(sigma_2) or np.isnan(sigma_1) or np.isnan(sigma_2):
```

Comment:
Can we delete this commented block of code?

Reply:
Done. As well as the other one in iter_5.